### PR TITLE
Eliminate B4B

### PR DESCRIPTION
**Pull request by johnmauff**
This commit eliminates a number of code blocks that are designed to produce B4B answers between the CPU and GPU versions of the code.  In particular, it focuses on code blocks that contain math intrinsic functions that have previously caused numerical differences.  Recently a flag has been added to the NVHPC compiler 'math_uniform' that provides a math intrinsic function calculations on the GPU that reproduces the CPU version.

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -78,9 +78,9 @@ ifeq (${FC},nvfortran)
     CPP          += cpp -C -P -traditional -Wno-invalid-pp-token -ffreestanding
     ifeq (${DEBUG_L},true)
          OPTS    += -g -O0
-         DM      += -D_B4B
+         # DM      += -D_B4B
```
Comment (sjsprecious) [inline src/Makefile]: just remove this line

File: src/Makefile [hunk 2]
```diff
@@ -124,7 +124,7 @@ ifeq (${FC},pgf90)
         ifeq (${DEBUG_L},true)
             OPTS += -acc -gpu=cc70,lineinfo,nofma,autocompare,math_uniform -Mnofma
         else
-            OPTS += -acc -gpu=cc70,lineinfo,math_uniform -Mpcast -Minfo=accel
+            OPTS += -acc -gpu=cc70,lineinfo,math_uniform -Mpcast -Minfo=accel -Mnofma
```
Comment (sjsprecious) [inline src/Makefile]: missing `nofma` for `-gpu` option

**Review by sjsprecious — CHANGES_REQUESTED** [review]

File: src/Makefile (hunk 1 above)
Comment (johnmauff) [reply src/Makefile]: I am fine with this change

File: src/Makefile (hunk 2 above)
Comment (johnmauff) [reply src/Makefile]: Thanks for the catch

**Review by sherimickelson — APPROVED** [review]

File: src/Makefile (hunk 1 above)
Comment (johnmauff) [reply src/Makefile]: Sorry about the confusion.  I think I have addressed your issue noew.

File: src/Makefile (hunk 2 above)
Comment (johnmauff) [reply src/Makefile]: This should be fixed now

File: src/Makefile [hunk 3]
```diff
@@ -74,13 +74,13 @@ endif
 #  NVHPC compiler
 #-----------------------------------------------------------------------------
 ifeq (${FC},nvfortran)
-    OPTS         += -Mfree -Ktrap=none -Mautoinline -Minline=reshape -Kieee
```
Comment (sjsprecious) [inline src/Makefile:77]: Add back the "-Kieee" flag

File: src/Makefile [hunk 4]
```diff
@@ -104,7 +104,7 @@ endif
 #  PGI compiler
 #-----------------------------------------------------------------------------
 ifeq (${FC},pgf90)
-    OPTS         += -Mfree -Ktrap=none -Mautoinline -Minline=reshape -Kieee
```
Comment (sjsprecious) [inline src/Makefile:107]: Add back the "-Kieee" flag

File: src/bc.F [hunk 5]
```diff
@@ -3409,209 +3409,193 @@ subroutine bcs_GPU(s)
       real, dimension(ib:ie,jb:je,kb:ke) :: s
 
       integer i,j,k
+      !$acc declare present(s)
 
-      !$acc data present(s)
```
Comment (sjsprecious) [inline src/bc.F:3413]: All my changes to `bc.F` are reverted. I would suggest to checking out my version and replacing your existing one.

File: src/bcast.F [hunk 6]
```diff
@@ -5,12 +5,12 @@ MODULE bcast_module
 
 interface bcast_int 
    module procedure bcast_int_scalar 
-   module procedure bcast_int_vector
```
Comment (sjsprecious) [inline src/bcast.F:8]: Same as `bc.F`. Use my version to replace your existing one.

File: src/cm1.F [hunk 7]
```diff
@@ -754,7 +754,6 @@ program cm1
 !-----------------------------------------------
```
Comment (sjsprecious) [inline src/cm1.F:754]: Same as `bc.F`. Use my version to replace your existing one but keep some changes you made.

File: src/comm.F [hunk 8]
```diff
@@ -13936,9 +13936,8 @@ subroutine prepcorners3_GPU(s,sw31,sw32,se31,se32,ss31,ss32,sn31,sn32,  &
       integer, intent(in) :: comm
       logical, parameter :: Debug = .FALSE.
       integer :: i,j
-      !$acc data present (s,sw31,sw32,se31,se32,ss31,ss32,sn31,sn32, &
```
Comment (sjsprecious) [inline src/comm.F:13939]: Same as `bc.F`. Use my version to replace your existing one.

File: src/constants.F [hunk 9]
```diff
@@ -8,12 +8,10 @@ MODULE constants
             cpdcv,rovcp,rddcp,rddcv,rddrv,cvdrd,cpdrd,eps,reps,repsm1,       &
             cpt,cvt,pnum,xlv,lathv,xls,lvdcp,condc,cpl,cpi,lv1,lv2,ls1,ls2,  &
             rhow,c_e1,c_e2,c_s,rcs,earth_radius,govtwo,mw,ms,surften,ion,os,ru
-
```
Comment (sjsprecious) [inline src/constants.F:11]: Same as `bc.F`. Use my version to replace your existing one.

File: src/domaindiag.F [hunk 10]
```diff
@@ -1139,15 +1124,16 @@ subroutine domaindiag(nstep,mtime,nwritet,trecs,trecw,qname,qunit,dt,dosfcflx, &
     ziflag:  &
     IF( ( testcase.ge.1 .and. testcase.le.7 ) .or. testcase.eq.9 .or. testcase.eq.10 .or. testcase.eq.11 )THEN
 
+        !$acc parallel private(i,j,k,zi,tmax,dtdz) &
+        !$acc   reduction(min:zimin) &
+        !$acc   reduction(max:zimax)
         zimin =  1.0e30
         zimax = -1.0e30
-        !$acc parallel default(present) reduction(min:zimin) reduction(max:zimax)
```
Comment (sjsprecious) [inline src/domaindiag.F:1144]: Except the changes to BFB flag, please keep my other changes.

File: src/droplet.F [hunk 11]
```diff
@@ -102,63 +102,56 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
       integer :: tmp_flag
 #endif
```
Comment (sjsprecious) [inline src/droplet.F:104]: Same as `bc.F`. Use my version to replace your existing one.

File: src/init_terrain.F [hunk 12]
```diff
@@ -178,7 +178,6 @@ subroutine init_terrain(xh,uh,xf,uf,yh,vh,yf,vf,rds,sigma,rdsf,sigmaf,  &
 
         zt = maxz
         rzt = 1.0/maxz
-        !$acc update device (zt,rzt)
```
Comment (sjsprecious) [inline src/init_terrain.F:181]: Same as `bc.F`. Use my version to replace your existing one.

File: src/input.F [hunk 13]
```diff
@@ -160,18 +160,23 @@ MODULE input
               bl_mynn_cloudmix,bl_mynn_mixqt,initflag,spp_pbl,bl_mynn_output, &
               nutk,nvtk,nwtk
 
-!$acc declare create(nqs1,nqs2,nqr,nqi,nqs,nqg,nnci,nncc, &
```
Comment (sjsprecious) [inline src/input.F:163]: Same as `bc.F`. Use my version to replace your existing one.

File: src/misclibs.F [hunk 14]
```diff
@@ -3252,21 +3244,24 @@ subroutine calcksquick(dt,uh,vh,mf,kmh,kmv,khh,khv,reqk)
         stop 1112
       endif
 
+      !$acc parallel default(present) private(k) reduction(max:fmax)
       fmax=-99999999.
-      !$acc parallel default(present) reduction(max:fmax)
```
Comment (sjsprecious) [inline src/misclibs.F:3256]: Except the changes to BFB flags, please keep my other changes.

File: src/param.F [hunk 15]
```diff
@@ -5334,7 +5334,6 @@ subroutine param(namelist_input,                                  &
 
         nvdiag = nvdiag+1
         vd_vadv = nvdiag
-        !$acc update device (vd_vadv)
```
Comment (sjsprecious) [inline src/param.F:5337]: Same as `bc.F`. Use my version to replace your existing one.

File: src/sfcphys.F [hunk 16]
```diff
@@ -50,9 +42,12 @@ subroutine getcecd(xh,u1,v1,s1,za,u10,v10,s10,xland,znt,ust,cd,ch,cq,avgsfcu,avg
       real :: wsp,wlast,var,rznt,usave,vsave,ssave
       logical :: convergeFail
 
-      !$acc declare present(xh,za,xland,u1,v1,s1,u10,v10,s10, &
```
Comment (sjsprecious) [inline src/sfcphys.F:53]: Except the changes to BFB flags, please keep my other changes.

File: src/turb.F [hunk 17]
```diff
@@ -2558,13 +2524,11 @@ subroutine tkekm(nstep,rtime,dt,ruh,rvh,rmh,zf,mf,rmf,zntmp,ust,ustt,rf,cmz,csz,
       enddo
       enddo
       enddo
-#ifdef _B4B03F
-      !$acc update device(grdscl,rgrdscl,lenscl,tk,dissten,kmh,kmv,khh,khv)
-#endif
 
     ENDIF
      
-    !$acc update host (grdscl)
```
Comment (sjsprecious) [inline src/turb.F:2567]: Except the changes to BFB flags, please keep my other changes.

**Review by sjsprecious — CHANGES_REQUESTED** [review]

**Comment by johnmauff** [comment]
There are some significant flaws in this pull request. I am submitting a new pull request that hopefully will not be as flawed.

**Pull request closed without merging**